common-lib/src/main/java/com/google/gct/idea/BasePluginInfoService.java

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright (C) 2015 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.gct.idea;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.gct.idea.flags.FlagReader;
+import com.google.gct.idea.flags.PropertiesFileFlagReader;
+import com.google.gct.idea.util.IntelliJPlatform;
+
+import com.intellij.ide.plugins.IdeaPluginDescriptor;
+import com.intellij.ide.plugins.PluginManager;
+import com.intellij.openapi.application.ApplicationInfo;
+import com.intellij.openapi.diagnostic.Logger;
+import com.intellij.openapi.extensions.PluginId;
+import com.intellij.util.PlatformUtils;
+
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * Subclasses of this class will inherit a full implementation of the {@link PluginInfoService}.
+ */
+// Don't expose PluginId in this service's API as it has a private
+// constructor and makes testing impossible.
+public abstract class BasePluginInfoService implements PluginInfoService {
+
+  private final String userAgent;
+  private final IdeaPluginDescriptor plugin;
+  private final FlagReader flagReader;
+
+  protected BasePluginInfoService(@NotNull String pluginUserAgentName, @NotNull String pluginId) {
+    this(
+        pluginUserAgentName,
+        PluginManager.getPlugin(PluginId.getId(pluginId)),
+        new PropertiesFileFlagReader());
+  }
+
+  @VisibleForTesting
+  BasePluginInfoService(
+      @NotNull String pluginUserAgentName,
+      @NotNull IdeaPluginDescriptor plugin,
+      @NotNull FlagReader flagReader) {
+    this.plugin = plugin;
+    this.userAgent = constructUserAgent(pluginUserAgentName, plugin.getVersion());
+    this.flagReader = flagReader;
+  }
+
+  @Override
+  public String getUserAgent() {
+    return userAgent;
+  }
+
+  @Override
+  public String getPluginVersion() {
+    return plugin.getVersion();
+  }
+
+  @Override
+  public String getPluginId() {
+    return plugin.getPluginId().getIdString();
+  }
+
+  @Override
+  public boolean shouldEnableErrorFeedbackReporting() {
+    return !getCurrentPlatform().equals(IntelliJPlatform.ANDROID_STUDIO);
+  }
+
+  /**
+   * !!!Please update {@link Feature}'s class documentation if you change the algorithm that
+   * determines whether a feature is enabled.!!!
+   */
```

Comment:
Should not the documentation of the algorithm be placed here and provide a link from Feature's class javadoc to this method? I think having the algorithm and the doc at the same place makes it easier keep them consistent.